msprime/species_tree.py

```diff
@@ -0,0 +1,834 @@
+        err += 'in the tree string!'
+        assert inlines[0].count('(') == inlines[0].count(')'), err
+        tree_string_raw = inlines[0]
+        tree_patterns = re.search('\\(.+\\)', tree_string_raw)
+        tree_string = tree_patterns.group(0)
+
+    # Make sure a tree string is found.
+    assert tree_string, 'No tree could be found in file {}.'.format(filename)
+
+    # Make sure that between the beginning and the end of the tree string,
+    # there are always more parentheses opened than closed.
+    parentheses_open = 0
+    for char in tree_string[:-2]:
+        if char == "(":
+            parentheses_open += 1
+        elif char == ")":
+            parentheses_open -= 1
+        err = 'The tree string in file {} '.format(filename)
+        err += 'appears to be malformed (possibly the first and '
+        err += 'last parentheses are missing).'
+        assert parentheses_open > 0, err
+
+    # Now that we know if the tree is in starbeast format or not, make
+    # sure that Ne is specified if and only if the tree is not in
+    # starbeast format.
+    if starbeast_format:
+        err = 'With species trees in StarBEAST format, Ne should not be specified.'
+        assert Ne is None, err
+    else:
+        assert Ne, 'Ne should be specified.'
+
+    # Use the tree string to generate a tree object.
+    species_tree = Tree(tree_string)
+
+    # Parse the newick tree string.
+    if starbeast_format:
+        species_tree.parse_extended_newick_string()
+    else:
+        species_tree.parse_newick_string()
+
+    # Set the extant progeny ids for each branch.
+    species_tree.set_extant_progeny_ids()
+
+    # Set the origin and termination times of all branches.
+    species_tree.set_times()
+
+    # Get a list of species IDs along with population sizes.
+    species_ids = []
+    terminal_pop_sizes = []
+    for edge in species_tree.get_edges():
+        if edge.get_is_extant():
+            species_ids.append(edge.get_node_ids()[1])
+            if starbeast_format:
+                edge_ne = edge.get_pop_size(generations_per_branch_length_unit)
+                terminal_pop_sizes.append(edge_ne)
+            else:
+                terminal_pop_sizes.append(Ne)
+    lists_are_sorted = False
+    while lists_are_sorted is False:
+        lists_are_sorted = True
+        for x in range(len(species_ids)-1):
+            if species_ids[x] > species_ids[x+1]:
+                species_ids[x], species_ids[x+1] = species_ids[x+1], species_ids[x]
+                terminal_pop_sizes[x], terminal_pop_sizes[x+1] = \
+                    terminal_pop_sizes[x+1], terminal_pop_sizes[x]
+                lists_are_sorted = False
+                break
+
+    # Determine at which time which populations should merge.
+    indices1 = []
+    indices2 = []
+    internal_pop_sizes = []
+    divergence_times = []
+    for edge in species_tree.get_edges():
+        if edge.get_is_extant() is False:
```

Comment:
It's not a good idea to compare booleans with object identity. 
```python
if edge.is_extant:
```
is more idiomatic.

Reply:
Thanks, this is now changed.